### PR TITLE
pilot: fix Gateway API conversion MatchExpressions handling for NotIn and DoesNotExist operators

### LINKED ISSUE #48044: Properly handle `NotIn`&`DoesNotExist` LabelSelectorOperators for namespaceReferences in k8s api gw -> istio gw conversion
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Conversion of K8S Gateway API to Istio API 

K8S Gateway API allows to set `AllowedRoutes` for the GW listeners to set the allowed namespace for the routes (via the `Selector` type)
https://gateway-api.sigs.k8s.io/concepts/api-overview/#referencing-gateways

During the conversion to the Istio API this field is supported. And the namespace label keys are taken to later to track the namespaces with those labels.

The problem comes with `NotIn`&`DoesNotExist` LabelSelectorOperators. These operators have the reverse logic. The namespaces with those MUST NOT be tracked and all other namespaces without these labels MUST be tracked.

https://github.com/istio/istio/blob/b69670715d924cfb03ea8c3cc5a31bd4edb36445/pilot/pkg/config/kube/gateway/conversion.go#L2249

cc @howardjohn  

### Version

```Text
Current master branch

https://github.com/istio/istio/blob/b69670715d924cfb03ea8c3cc5a31bd4edb36445/pilot/pkg/config/kube/gateway/conversion.go#L2249
```

### Additional Information

_No response_

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### PR DESCRIPTION
**Pull request by mikemorris**
**Please provide a description of this PR:**

Fixes #48044 

Added a few additional listeners in the test fixture YAML to add coverage for AllowedRoutes selector MatchExpressions, but think I missed updating a bit of boilerplate to get the golden files regenerating properly (and I might need to add additional routes to target these new listeners?) - I'm currently hitting the following error when running `REFRESH_GOLDEN=1 go test ./pilot/pkg/config/kube/gateway -run TestConvertResources/route-binding -v`

```
    diff.go:111: Refreshing golden file testdata/route-binding.yaml.golden
    conversion_test.go:1229: failed to parse CRD: configuration is invalid: 2 errors occurred:
                * config namespace and dnsName in host entry cannot be empty
                * invalid namespace value "" in gateway
```

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 1]
```diff
@@ -2247,6 +2247,10 @@ func getNamespaceLabelReferences(routes *k8s.AllowedRoutes) []string {
 		res = append(res, k)
 	}
 	for _, me := range routes.Namespaces.Selector.MatchExpressions {
+		if me.Operator == "NotIn" || me.Operator == "DoesNotExist" {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/conversion.go:2250]: nit: use LabelSelectorOpDoesNotExist from metav1

**Review by howardjohn — APPROVED** [review]
LGTM. A comment along the lines of "over-matching is fine, we just cannot under match" would help explain this. But looks good

File: pilot/pkg/config/kube/gateway/testdata/route-binding.status.yaml.golden [hunk 2]
```diff
@@ -247,7 +247,7 @@ status:
     parentRef:
       name: gateway
       namespace: istio-system
-      sectionName: namespace-selector
+      sectionName: selector-labels
```
Comment (MorrisLaw) [inline pilot/pkg/config/kube/gateway/testdata/route-binding.status.yaml.golden:250]: nit: looks like extra whitespace on this line, right after labels

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 3]
```diff
@@ -2247,6 +2247,10 @@ func getNamespaceLabelReferences(routes *k8s.AllowedRoutes) []string {
...
+			res = append(res, "*")
```
[the remark this answers is not in the document]
Comment (mikemorris) [reply pilot/pkg/config/kube/gateway/conversion.go:2252]: The `.Intersection` method inside `namespaceEvent` implements this.

https://github.com/istio/istio/blob/64d8e69aa62e27068312f9b7006450aaaa4aa336/pilot/pkg/config/kube/gateway/controller.go#L312-L335

[the remark this answers is not in the document]
Comment (mikemorris) [reply pilot/pkg/config/kube/gateway/conversion.go:2252]: https://github.com/istio/istio/blob/master/pkg/config/host/names.go#L75-L99

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 4]
```diff
@@ -2414,7 +2423,7 @@ func parentRefString(ref k8s.ParentReference) string {
 		ptr.OrEmpty(ref.Namespace))
 }
 
-// buildHostnameMatch generates a VirtualService.spec.hosts section from a listener
+// buildHostnameMatch generates a Gateway.spec.servers.hosts section from a listener
```
Comment (mikemorris) [inline pilot/pkg/config/kube/gateway/conversion.go:2426]: This comment appears to have been inaccurate - the VirtualService.spec.hosts section would be built from HTTPRoute.spec.hostnames - [Listener](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Listener) configuration is used to build the Istio Gateway servers.

File: releasenotes/notes/gateway-allowedroutes-fix.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: networking
```
Comment (mikemorris) [inline releasenotes/notes/gateway-allowedroutes-fix.yaml:3]:
```suggestion
area: traffic-management
```

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 6]
```diff
@@ -2424,7 +2433,10 @@ func buildHostnameMatch(localNamespace string, r GatewayResources, l k8s.Listene
 
 	resp := []string{}
 	for _, ns := range namespacesFromSelector(localNamespace, r, l.AllowedRoutes) {
-		resp = append(resp, fmt.Sprintf("%s/%s", ns, hostname))
+		// This check is necessary to prevent adding a hostname with an invalid empty namespace
+		if len(ns) > 0 {
+			resp = append(resp, fmt.Sprintf("%s/%s", ns, hostname))
+		}
```
Comment (mikemorris) [inline pilot/pkg/config/kube/gateway/conversion.go:2439]: Excluding the `""` empty string namespace picked up by tracking `"*"` when building the Istio Gateway server hosts array was the trick needed to fix the invalid generated CRDs.

File: pkg/config/validation/validation.go [hunk 7]
```diff
@@ -1075,19 +1075,10 @@ func validateNamespaceSlashWildcardHostname(hostname string, isGateway bool) (er
...
-	} else {
-		// namespace can be * or . or a valid DNS label in gateways
-		if parts[0] != "*" && parts[0] != "." {
-			if !labels.IsDNS1123Label(parts[0]) {
-				errs = appendErrors(errs, fmt.Errorf("invalid namespace value %q in gateway", parts[0]))
-			}
+	// namespace can be * or . or ~ or a valid DNS label in sidecars or gateways
+	if parts[0] != "*" && parts[0] != "." && parts[0] != "~" {
+		if !labels.IsDNS1123Label(parts[0]) {
+			errs = appendErrors(errs, fmt.Errorf("invalid namespace value %q", parts[0]))
```
Comment (mikemorris) [inline pkg/config/validation/validation.go:1081]: `~` namespace labels _are_ used by Istio Gateways when converting from Gateway API gateways and no matching routes can be found - I believe these generated resources were incorrectly failing validation before, see https://github.com/istio/istio/blob/e46a62df82104afdf4006775c344e690554cd7cc/pilot/pkg/config/kube/gateway/conversion.go#L2442-L2447

File: pilot/pkg/config/kube/gateway/testdata/route-binding.status.yaml.golden [hunk 8]
```diff
@@ -164,7 +164,287 @@ status:
...
+      message: No errors found
+      reason: Programmed
+      status: "True"
+      type: Programmed
+    - lastTransitionTime: fake
+      message: No errors found
+      reason: ResolvedRefs
+      status: "True"
+      type: ResolvedRefs
+    name: slctr-expr-notin-no
```
Comment (mikemorris) [inline pilot/pkg/config/kube/gateway/testdata/route-binding.status.yaml.golden:279]: This is okay, the `group-namespace1` route is excluded but the `default/bind-all` route is not.

**Comment by mikemorris** [comment]
I ended up touching a few additional spots of logic that seemed incorrect (left comments highlighting them), this should be complete now but could use another set of eyes.

/retest

File: pkg/config/validation/validation.go (hunk 7 above)
Comment (howardjohn) [reply pkg/config/validation/validation.go:1081]: The internal generate ones don't actually go through validation in real world -- only in the tests. 

Not sure how much we care about maintaining that property. Looks like in other parts we allowlist these "internal gateways" (see code like `gatewaySemantics := cfg.Annotations[constants.InternalRouteSemantics] == constants.RouteSemanticsGateway`)

Comment (mikemorris) [reply pkg/config/validation/validation.go]: Rewrote this logic to plumb through a new `internal.istio.io/gateway-semantics` annotation, see latest commit.

File: releasenotes/notes/gateway-allowedroutes-fix.yaml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - 48044
+
+releaseNotes:
+- |
+  **Fixed** Gateway API AllowedRoutes handling for NotIn and DoesNotExist LabelSelector MatchExpressions.
```
Comment (dhawton) [inline releasenotes/notes/gateway-allowedroutes-fix.yaml:9]:
```suggestion
  **Fixed** Gateway API `AllowedRoutes` handling for `NotIn` and `DoesNotExist` label selector match expressions.
```

**Pull request merged**